src/http-api.js

```diff
@@ -83,6 +83,44 @@ module.exports.MatrixHttpApi = function MatrixHttpApi(event_emitter, opts) {
     this.uploads = [];
 };
 
+/**
+ * Construct a Timeout Context.
+ * Set onProgressCheck to receive progressEvent from an XMLHTTPRequest.
+ * TimeoutContext will debounce the timeoutFunction for waitMs if progress
+ * is being made.
+ *
+ * @constructor
+ * @param {Function} timeoutFunction The function to debounce with waitMs.
+ * @param {Number} waitMs The time to debounce the callback in miliseconds.
+ *
+ */
+const TimeoutContext = function TimeoutContext(timeoutFunction, waitMs) {
+    this.lastLoaded = null;
+    this.jobName = null;
+    this.timeoutFunction = timeoutFunction;
+    this.waitMs = waitMs;
+};
+
+// For testing
+module.exports.__TimeoutContext = TimeoutContext;
+
+TimeoutContext.prototype.onProgressCheck = function(progressEvent) {
+    if (this.lastLoaded || progressEvent.loaded > this.lastLoaded) {
```

Comment:
Should be `if (!this.lastLoaded ||` or rather `if (this.lastLoaded != null`. Is it safe to just set loaded to 0, the first onProgressEvent may have loaded `0` or is that not the case?